Add Capella & Deneb light client support

eserilev opened the pull request:
## Issue Addressed

#4896

## Proposed Changes

conform to capella/deneb light client specs

- [Capella light client spec](https://github.com/ethereum/consensus-specs/tree/dev/specs/capella/light-client)
- [Deneb light client spec](https://github.com/ethereum/consensus-specs/tree/dev/specs/deneb/light-client)

## Additional Info

https://github.com/sigp/lighthouse/pull/3954 is a blocker

GeemoCandama committed:
rebase and add comment

GeemoCandama committed:
conditional test

GeemoCandama committed:
test

GeemoCandama committed:
optimistic chould be working now

GeemoCandama committed:
finality should be working now

GeemoCandama committed:
try again

GeemoCandama committed:
try again

GeemoCandama committed:
clippy fix

GeemoCandama committed:
add lc bootstrap beacon api

GeemoCandama committed:
add lc optimistic/finality update to events

GeemoCandama committed:
fmt

GeemoCandama committed:
That error isn't occuring on my computer but I think this should fix it

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

# Conflicts:
#	beacon_node/beacon_chain/src/events.rs
#	beacon_node/http_api/src/lib.rs
#	beacon_node/http_api/src/test_utils.rs
#	beacon_node/http_api/tests/main.rs
#	beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
#	beacon_node/lighthouse_network/src/rpc/methods.rs
#	beacon_node/lighthouse_network/src/service/api_types.rs
#	beacon_node/network/src/beacon_processor/worker/rpc_methods.rs
#	beacon_node/tests/test.rs
#	common/eth2/src/types.rs
#	lighthouse/src/main.rs

jimmygchen committed:
Add missing test file

jimmygchen committed:
Update light client types to comply with Altair light client spec.

jimmygchen committed:
Fix test compilation

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

jimmygchen committed:
Support deserializing light client structures for the Bellatrix fork

jimmygchen committed:
Move `get_light_client_bootstrap` logic to `BeaconChain`. `LightClientBootstrap` API to return `ForkVersionedResponse`.

jimmygchen committed:
Misc fixes.
- log cleanup
- move http_api config mutation to `config::get_config` for consistency
- fix light client API responses

jimmygchen committed:
Add light client bootstrap API test and fix existing ones.

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

jimmygchen committed:
Fix test for `light-client-server` http api config.

jimmygchen committed:
Appease clippy

jimmygchen committed:
Add Altair light client SSZ tests

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into light_client_beacon_api_1

eserilev committed:
updates to light client header

eserilev committed:
light client header from signed beacon block

eserilev committed:
using options

eserilev committed:
implement helper functions

eserilev committed:
placeholder conversion from vec hash256 to exec branch

eserilev committed:
add deneb

eserilev commented:
Here's a quick update on what I've got so far:

- I explored superstruct usage within `LightClientHeader` but I believe using `Option` and handling the different execution payload types (i.e. capella, deneb) to be the better solution here. If we were to use a superstruct I think we'd need to make some changes to additional light client related structs which feels bad.
- I'm a bit iffy on the code I wrote that creates the merkle proof. This is the main thing I'm trying to figure out at the moment
- I'm wondering what kind of tests we should write here.
- I need to get SSZ encode and decode to play nice with `Option<ExecutionPayloadHeader>`

eserilev committed:
using fixed vector

eserilev committed:
remove unwraps

eserilev committed:
by epoch

eserilev committed:
compute merkle proof

eserilev committed:
merkle proof

eserilev committed:
update comments

eserilev committed:
resolve merge conflicts

eserilev commented on consensus/types/src/light_client_header.rs:
```diff
@@ -15,12 +21,160 @@ use test_random_derive::TestRandom;
     pub beacon: BeaconBlockHeader,
+    #[test_random(default)]
+    #[ssz(skip_serializing, skip_deserializing)]
+    pub execution: Option<ExecutionPayloadHeader<E>>,
```
im skipping serializing here, because Option is causing errors. Theres probably a way to fix this, I just haven't figured it out yet

dapplion replied:
Can you try to use superstruct here? Since fields are all fork dependant it fits perfectly to do the job. At least give it a try on another branch and report here which blockers are you encountering.

jimmygchen replied:
Yeah i think we should try use `superstruct` to ensure we're dealing with the right types on the each fork and it would also be more convenient when trying to figure out which fork variant it is.

eserilev replied:
yep np, i'll go ahead and use superstruct

eserilev committed:
linting

eserilev commented:
I opened the PR, but theres a few TODOs ill be taking care of tomorrow

- add tests for beacon block body merkle proof
- add tests for new light client header helper functions

also a question:

the new helper functions are considered dead code at the moment which causes linting errors. I think I can tell clippy to ignore the dead code, but should these helper functions be called anywhere? (outside of tests)

eserilev commented on consensus/types/src/light_client_header.rs:
```diff
@@ -15,12 +21,160 @@ use test_random_derive::TestRandom;
 }
+
+impl<E: EthSpec> LightClientHeader<E> {
+    fn new(chain_spec: ChainSpec, block: SignedBeaconBlock<E>) -> Result<Self, Error> {
```
`ChainSpec` is needed here to figure out what fork were at. Unless theres a better way to do it?

jimmygchen replied:
Depends on whether we need to do a fork consistency check here - if we don't, we could use `block.fork_name_unchecked` (but I haven't read the whole thing yet)

jimmygchen commented on consensus/types/src/beacon_block_body.rs:
```diff
@@ -508,6 +518,60 @@ impl<E: EthSpec> From<BeaconBlockBody<E, FullPayload<E>>>
 }
 
+impl<T: EthSpec> BeaconBlockBody<T> {
+    pub fn compute_merkle_proof(
```
FYI - there's something similar (computing merkle proof) going on with blobs inclusion proof - it's not completely identical but there's probably some parts that are reusable - maybe worth renaming this function at least to avoid confusions 
https://github.com/sigp/lighthouse/pull/4900/files#diff-93ee4b8bd96daedfce7931df6ed17fcfe4dd8e9f75a8681fc0613caf71c55276R112-R180

eserilev replied:
Renamed to `block_body_merkle_proof`.  

re: reusable code: 
In `kzg_commitment_merkle_proof` lines 158-180 can be replaced by `block_body_merkle_proof`. Just need to figure out how to make that function available within `BeaconBlockBodyRef<'a, T, Payload>` impl and `BeaconBlockBody` impl

jimmygchen commented:
Re the unused helper functions, I guess you meant `LightClientHeader::new` (equivalent to the spec's [`block_to_light_client_header`](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#block_to_light_client_header) function, may be worth adding a comment on top of the function). This will be used when deriving light client data, see [deriving light client data](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#deriving-light-client-data) in altair specs.

@dapplion is currently working on computing and caching the light client data for Altair (#4926) , this function would be used when producing the light client data, I think your function implementation would replace this:
https://github.com/sigp/lighthouse/compare/unstable...dapplion:lighthouse:lc-prod-recent-updates?expand=1#diff-19cf8a8a12dc28ae7de4622c52ec606add9dbe2abef58b59209cccf372975234R219-R224

jimmygchen committed:
Merge branch 'unstable' into light-client-ssz-tests

# Conflicts:
#	beacon_node/beacon_chain/src/beacon_chain.rs
#	consensus/types/src/light_client_bootstrap.rs
#	consensus/types/src/light_client_header.rs

jimmygchen commented:
FYI I created a PR that runs light client SSZ test vectors from `consensus-spec-tests` - if it's useful for testing your SSZ type changes, feel free to merge it into your branch. Otherwise we can review that one after your PR is merged.
https://github.com/sigp/lighthouse/pull/4938

eserilev committed:
superstruct attempt

eserilev committed:
superstruct changes

eserilev committed:
lint

eserilev committed:
altair

eserilev committed:
update

eserilev committed:
update

eserilev commented:
I pushed up my changes regarding superstruct. Theres still a small issue or two that needs to be resolved. To summarize:

## Superstruct Changes

- `LightClientHeader` is now a superstruct with three variants, `Altair`, `Capella` and `Deneb`. Each variant has its own implementation for `block_to_light_client_header`, `get_lc_execution_root`, `is_valid_light_client_header`

#### `LightClientBootstrap::from_beacon_state`

- Added `ChainSpec` as one of the function arguments so we can make a fork consistency check here. 
- The function checks what fork we're on (altair, capella or deneb) and chooses to construct the correct `LightClientHeader` variant based on the fork check. There might be a better way to do this.

#### `LightClientFinalityUpdate::new`
- added an additional function argument `attested_block: SignedBeaconBlock<T>` which is required to create the `attested_header`        
- updated `finalized_block` arg to `&SignedBeaconBlock<T>` (from `SignedBlindedBeaconBlock`), beacause we need the execution payload in order to construct the `LightClientHeader`

#### `LightClientOptimisticUpdate::new`

- added an additional function argument `attested_block: SignedBeaconBlock<T>` which is required to create the `attested_header`
- The function checks what fork we're on (altair, capella or deneb) and chooses to construct the correct `LightClientHeader` variant based on the fork check. There might be a better way to do this.


#### `LightClientUpdate::new`
- added `attested_block: SignedBeaconBlock<T>` as an argument to `new`, I believe this is required per the [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#create_light_client_update) 
- The function checks what fork we're on (altair, capella or deneb) and chooses to construct the correct `LightClientHeader` variant based on the fork check. There might be a better way to do this.


#### TestRandom

- TestRandom cannot be derived for enum types. With the superstruct changes to `LightClientHeader` deriving `TestRandom` for `LightClientFinalityUpdate`, `LightClientOptimisticUpdate`, `LightClientUpdate` and `LightClientBootstrap` causes compilation issues. I'll need to dig in here to resolve


I also left a comment regarding a small change in `beacon_node/lighthouse_network/src/rpc/methods.rs` thats causing an issue

## Fetching the full beacon block

In Altair, we only needed the block header to construct a `LightClientHeader`. In Capella and Deneb we need the execution payload as well. In places where we were fetching a `SignedBlindedBeaconBlock` we are now required to fetch the full beacon block. Fetching the full beacon block is done by calling the async function `get_block`. Since the LightClient code is sync, I created a runtime object (via tokio::runtime) to allow for executing async code within a sync function. I'm not sure what the overhead is for creating these runtimes, or if theres a better/more standard way of doing this.

eserilev commented on beacon_node/lighthouse_network/src/rpc/methods.rs:
```diff
@@ -574,7 +574,7 @@ impl<T: EthSpec> std::fmt::Display for RPCResponse<T> {
                     f,
                     "LightClientBootstrap Slot: {}",
-                    bootstrap.header.beacon.slot
+                    bootstrap.header.beacon().slot
```
With the  `LightClientHeader` update to superstruct, `beacon` can only be accessed via method call (unless we convert `header` to a specific `LightClientHeader` variant type)

This causes linting issues further downstream 
e.g. 

`large size difference between variants`

```
pub enum RPCCodedResponse<T: EthSpec> {
425 | |     /// The response is a successful.
426 | |     Success(RPCResponse<T>),
    | |     ----------------------- the largest variant contains at least 552 bytes
427 | |
428 | |     Error(RPCResponseErrorCode, ErrorType),
    | |     -------------------------------------- the second-largest variant contains at least 25 bytes
...   |
431 | |     StreamTermination(ResponseTermination),
432 | | }
    | |_^ the entire enum is at least 552 bytes
    
    help: consider boxing the large fields to reduce the total size of the enum
    |
426 |     Success(Box<RPCResponse<T>>),
```

If I follow clippys suggestions and Box `RPCResponse` this forces me to make some larger scale changes downstream to make the compiler happy. I'm a bit hesitant to make changes here, maybe we should ignore clippys suggestions?

jimmygchen replied:
I think it's due to `RPCResponse::LightClientBootstrap` variant here:
https://github.com/sigp/lighthouse/blob/03063bef268ce56d85658313030d511caa49bc49/beacon_node/lighthouse_network/src/rpc/methods.rs#L393

An enum size is the largest value of all variants, so we'd usually avoid having large objects directly stored in enum.

I think can just `Arc` the `LightClientBootstrap<T>` object here, like how it's done for blocks and blobs, since we only need a reference later to construct the SSZ object here:

https://github.com/sigp/lighthouse/blob/8a599ec7dc978f8d2a49100e3d712794b56eba2e/beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs#L76

eserilev replied:
nice, that fixed it, thanks!

jimmygchen commented:
Awesome, I'll take a look later today or tomorrow!

eserilev committed:
changes to light_client_optimistic_ and finality

eserilev commented:
I've pushed some significant changes/fixes to resolve some issues that I caught after writing my latest update. I'll edit my post to reflect those changes

dapplion commented on beacon_node/beacon_chain/src/light_client_finality_update_verification.rs:
```diff
@@ -67,7 +67,8 @@ impl<T: BeaconChainTypes> VerifiedLightClientFinalityUpdate<T> {
         chain: &BeaconChain<T>,
         seen_timestamp: Duration,
     ) -> Result<Self, Error> {
-        let gossiped_finality_slot = light_client_finality_update.finalized_header.beacon.slot;
```
Note that having to compute the full LightClientFinality update every time verification is called is not the optimal final strategy. PR https://github.com/sigp/lighthouse/pull/4969 already switches to computing this updates eagerly in another thread and using the cached values here on verification.

So I'm not sure if it's just better to merge https://github.com/sigp/lighthouse/pull/4969 first so we don't have to deal with the runtime here.

eserilev replied:
Yeah I also think it makes sense to wait to merge #4969 first

dapplion commented on consensus/types/src/light_client_bootstrap.rs:
```diff
@@ -1,46 +1,63 @@
+        chain_spec: &ChainSpec,
+        beacon_state: &mut BeaconState<T>,
+        block: SignedBeaconBlock<T>,
+    ) -> Result<Self, Error> {
```
It is unlikely that we will ever create bootstrap messages from the BeaconState in practice. Similar to my comment above, the existence of these functions come from a literal implementation of the spec, but may not survive latter revisions. We can merge this code for now, but let's keep it mind that it may be deleted soon.

jimmygchen replied:
I'm curious if you have any optimisation approach for this in mind? if not, I'll just create an issue and add it to the [tracking issue](https://github.com/sigp/lighthouse/issues/3651) to make sure that we don't miss this before turning light client server on.

dapplion replied:
Yes see https://hackmd.io/@dapplion/BkxPbc6eQT

dapplion commented on consensus/types/src/light_client_update.rs:
```diff
@@ -120,22 +119,67 @@ impl<T: EthSpec> LightClientUpdate<T> {
+                });
+            }
+        };
+
```
I would turn that into a match to the Fork enum, and return a tuple (attested_header, finalized_header)

dapplion commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -6455,13 +6455,18 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            self
+                .get_block(block_root).await
+            }
+        )? else {
```
Note that .get_block() fetches an execution payload from the EL, which may timeout blocking this task for maybe seconds? get_light_client_bootstrap is called by API and ReqResp:
- API spawns a blocking task, so it's ok to block

https://github.com/sigp/lighthouse/blob/31044402ee180ff937027ec842513bef90d7eec8/beacon_node/http_api/src/lib.rs#L2421-L2422

- ReqResp spawns a blocking fn, so ok to block too

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/beacon_processor/src/lib.rs#L1413-L1414

Having to wait for the EL should only degrade the experience for the caller. This is ok since this penalty already exist for queries on the API and ReqResp that involves full blocks

eserilev committed:
merge unstable

eserilev committed:
refactor

ec2 commented:
Hey friends! Very excited for this PR to land. Currently we need to maintain our own structs and would like to instead use Lighthouse's fork/version aware stuff. https://github.com/ralexstokes/ethereum-consensus/ also similarly has not implemented these types yet and also currently doesn't support merkle proofs. 

Light client server functionality aside, it would be dope if just the struct def'n land so that we can make API calls to Lodestar to get the data we need. 

Is this PR active? Or will we need continue to maintain our own structs for the foreseeable future?

jimmygchen commented:
Hi @ec2,
Yes this PR is still active! We're currently reviewing #4969, and once that's merged we'll revisit this PR, hopefully won't be too long!

ec2 commented:
Thanks for the update @jimmygchen !!

jimmygchen commented:
@eserilev #4969 has been merged 🎉  so it should unblock this PR! 
There are now some conflicts that needs to be resolved.

eserilev committed:
resolved merge conflicts

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into capella_deneb_light_client_types

eserilev commented:
I've resolved the merge conflicts. Need to take another look through the code to make sure its ready for a review. Will ping here as soon as I think its ready.

On a separate note:
We decided to change `LightClientHeader` to a superstruct. The struct used to derive `TestRandom`, I believe for compatibility with the `ssz_test!` macro. With the changes to superstruct, `LightClientHeader` is now an enum. `TestRandom` doesnt work with enums. I've commented out the calls to `ssz_test!` for now. If anyone has any better ideas here let me know

jimmygchen commented:
Sounds good!

You can have `TestRandom` added to the superstruct variant attributes, so each generated struct derives it, like this:
https://github.com/sigp/lighthouse/blob/31044402ee180ff937027ec842513bef90d7eec8/consensus/types/src/beacon_block_body.rs#L25-L55

eserilev commented:
Sorry I should have been more clear. Yes I was able to add `TestRandom` to the superstruct itself within the variant attributes like you mentioned. However, this causes issues downstream e.g. `LightClientBootstrap`. Since one of the fields in `LightClientBootstrap` is a `LightClientHeader` I am unable to derive `TestRandom` for `LightClientBootstrap`. The issue I guess is that the variants of `LightClientHeader` derive `TestRandom`, but `LIghtClientHeader` itself does not.

eserilev committed:
block_to_light_client_header fork aware

eserilev committed:
fmt

eserilev commented on beacon_node/beacon_chain/src/light_client_server_cache.rs:
```diff
@@ -127,13 +135,13 @@ impl<T: BeaconChainTypes> LightClientServerCache<T> {
                     // light_client header becomes expensive
-                    attested_header: block_to_light_client_header(attested_block.message()),
-                    finalized_header: block_to_light_client_header(finalized_block.message()),
+                    attested_header: block_to_light_client_header(fork_name, attested_block)?,
+                    finalized_header: block_to_light_client_header(fork_name, finalized_block)?,
```
For both the `attested_header` and the `finalized_header` I'm getting the `fork_name` from the `attested_block`. This might be a stupid question, but I'm wondering if there could be fork boundary issues here where the `attested_block` is in a different fork than the `finalized_block`?

dapplion replied:
> This might be a stupid question, but I'm wondering if there could be fork boundary issues here where the attested_block is in a different fork than the finalized_block?

Definitely not a stupid question. We try to schedule forks such that the fork epoch is a multiple of 256, to keep historical vectors in the same fork. Indirectly that makes lightclient periods align with fork boundaries. However some other network, testnet etc could choose a fork boundary such that it does not align with the start of the sync committee period. For simplicity we can accept to crash for now.

ec2 replied:
I've thought about this before as well and forgot to ask someone. Thanks for this question @eserilev and thanks for the asnwer @dapplion

eserilev committed:
comment fix

eserilev committed:
comment fix

dapplion commented on beacon_node/beacon_chain/src/light_client_server_cache.rs:
```diff
@@ -239,18 +248,24 @@ fn is_latest_optimistic_update<T: EthSpec>(
+) -> Result<LightClientHeader<T>, BeaconChainError> {
+    let light_client_header = match fork_name {
+        ForkName::Base => return Err(LightClientError::AltairForkNotActive.into()),
+        ForkName::Merge => return Err(LightClientError::AltairForkNotActive.into()),
```
Fork Merge fork is post Altair, so should be active

dapplion commented on consensus/types/src/light_client_finality_update.rs:
```diff
@@ -70,9 +54,61 @@ impl<T: EthSpec> LightClientFinalityUpdate<T> {
+                        )?
+                        .into();
+
+                    (attested_header, finalized_header)
```
Nit, but this can be compacted by dropping the intermediary variables

eserilev committed:
include merge fork, update deserialize_by_fork, refactor

eserilev committed:
fmt

eserilev committed:
pass by ref to prevent clone

eserilev committed:
rename merkle proof fn

eserilev commented:
I think this PR is in a good spot for another review

dapplion reviewed (commented):
Looks final to me! Nit capitalize `LIGHTCLIENT_BOOTSTRAP` as `LIGHT_CLIENT_BOOTSTRAP`. Also could you add a TODO or FIXME for the commented tests?

eserilev committed:
add FIXME

dapplion reviewed (approved):
Good to merge, thank you!

eserilev committed:
LightClientHeader TestRandom

eserilev commented:
@jimmygchen 

commit [06da24a0f1b6f5a66dffe5da8611583a56f166a6](https://github.com/sigp/lighthouse/pull/4946/commits/06da24a0f1b6f5a66dffe5da8611583a56f166a6) is my attempt at a workaround for TestRandom on structs that include a `LightClientHeader` field. My reasoning here is that since `LightClientHeaderDeneb` is a superset of all other variants, it makes most sense to use it to implement `TestRandom` on `LightClientHeader`. Since `TestRandom` is strictly for unit tests, it felt ok for me to do this. The other option here is to convert `LightClientBootstrap`, `LightClientUpdate`, `LightClientFinalityUpdate` and `LightClientOptimisticUpdate` to superstructs. I'm hoping to avoid that if possible. Let me know what you think.

eserilev committed:
fix comments

ec2 commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,183 @@
+#[serde(bound = "E: EthSpec", deny_unknown_fields)]
+#[ssz(enum_behaviour = "union")]
+#[arbitrary(bound = "E: EthSpec")]
+pub struct LightClientHeader<E: EthSpec> {
```
I think `ForkVersionDeserialize` should be implemented for this struct

eserilev replied:
added `ForkVersionDeserialize`

eserilev committed:
fork version deserialize

eserilev committed:
merge unstable

jimmygchen commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -6617,13 +6618,14 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-            .get_blinded_block(block_root)?
-            .map(|block| (block.state_root(), block.slot()))
-        else {
+        let runtime = tokio::runtime::Runtime::new().map_err(|_| Error::RuntimeShutdown)?;
```
Could we reuse the `self.task_executor` to run a blocking task?

eserilev replied:
yes! changed in dde9274

jimmygchen commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -6633,12 +6635,12 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-            ForkName::Altair | ForkName::Merge => {
-                LightClientBootstrap::from_beacon_state(&mut state)
+            ForkName::Altair | ForkName::Merge | ForkName::Capella | ForkName::Deneb => {
+                LightClientBootstrap::from_beacon_state(&self.spec, &mut state, &block)
```
Could we move `&self.spec` to the last arg here? it's just by convention in Lighthouse - usually `spec` and `logger` are placed as last arguments of a function, for readability

jimmygchen commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -6617,13 +6618,14 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-        else {
+        let runtime = tokio::runtime::Runtime::new().map_err(|_| Error::RuntimeShutdown)?;
+
+        let Some(block) = runtime.block_on(async { self.get_block(block_root).await })? else {
```
Any reason why we need the full block here?

jimmygchen replied:
Oh right! We need the execution payload now

jimmygchen commented on beacon_node/beacon_chain/src/light_client_server_cache.rs:
```diff
@@ -72,6 +76,7 @@ impl<T: BeaconChainTypes> LightClientServerCache<T> {
     pub fn recompute_and_cache_updates(
         &self,
         log: &Logger,
```
nit: move `chain_spec` and `log` to last

jimmygchen commented on beacon_node/beacon_chain/src/light_client_server_cache.rs:
```diff
@@ -99,6 +105,9 @@ impl<T: BeaconChainTypes> LightClientServerCache<T> {
         let attested_slot = attested_block.slot();
 
+        let fork_name =
+            chain_spec.fork_name_at_epoch(attested_slot.epoch(T::EthSpec::slots_per_epoch()));
```
I think you can get fork name from the `attested_block.fork_name`? (or you could even just do it in `block_to_light_client_header`)

eserilev committed:
move fn arguments, fork name calc

eserilev committed:
use task executor

jimmygchen commented on consensus/types/src/light_client_finality_update.rs:
```diff
@@ -45,15 +47,9 @@ impl<T: EthSpec> LightClientFinalityUpdate<T> {
-        finalized_block: &SignedBlindedBeaconBlock<T>,
+        attested_block: &SignedBeaconBlock<T>,
+        finalized_block: &SignedBeaconBlock<T>,
     ) -> Result<Self, Error> {
```
Where is this function used? 
Looks like we now construct the updates from `LightClientServerCache`, so no longer need this?

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,208 @@
+        })
+    }
+
+    #[allow(dead_code)]
+    fn get_lc_execution_root(&self) -> Option<Hash256> {
+        None
+    }
+
+    #[allow(dead_code)]
+    fn is_valid_light_client_header(&self) -> Result<bool, Error> {
+        Ok(true)
+    }
```
Looks like these functions aren't required by full nodes and can be excluded?
Was the idea to include it for the users of the `consensus-types` crate? This might be outside the responsibility of Lighthouse though and not sure if we should maintain something we don't use. They're also private so not accessible anyway?

eserilev replied:
Yeah I added these because they were included in the spec. Wasn't really sure what our policy was here, but I can go ahead and delete

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,208 @@
+            ForkName::Deneb => serde_json::from_value(value)
+                .map(|light_client_header| Self::Deneb(light_client_header))
+                .map_err(serde::de::Error::custom),
+            ForkName::Base | ForkName::Merge => Err(serde::de::Error::custom(format!(
```
`ForkName::Merge` should fall under the `ForkName::Altair` arm

eserilev committed:
remove unneeded fns

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,143 @@
+    variants(Altair, Capella, Deneb),
+    variant_attributes(
+        derive(
+            Default,
```
Doesn't seem necessary to derive `Default` here?

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,143 @@
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, arbitrary::Arbitrary, PartialEq)]
+#[serde(untagged)]
+#[serde(bound = "E: EthSpec", deny_unknown_fields)]
+#[ssz(enum_behaviour = "union")]
```
Why do we use `union` here?  It adds a one byte variant selector, and i think we want to use `transparent` instead?
https://github.com/sigp/ethereum_ssz/blob/649ed02ef19c6b2c364d0f4b514cc1c14f6b44d0/ssz_derive/src/lib.rs#L9

eserilev replied:
I used `union` because the compiler was complaining about not being able to dervie `Decode` for `transparent` enum behaviours

jimmygchen replied:
I see, I think I've done that before too lol but the generated implementation won't be able to decode the light client data (SSZ) received from peers, as they aren't encoded with the one-byte variant selector.  In this case we'll need to decode manually without the `Decode` derive.

I think we'll need something similar to how we handle `BeaconBlock` received over gossip - using the fork digest from the topic to determine the fork name, an example topic looks like this: `/eth2/e1925f3b/beacon_block/ssz_snappy`, where `e1925f3b` is the fork digest and it tells us the fork of this topic, so we should be able to decode using this information.

For example:
https://github.com/sigp/lighthouse/blob/4273004bd905341f6a3aa3ecbe203ce4b66e6497/beacon_node/lighthouse_network/src/types/pubsub.rs#L173-L204

and the LightClient decoding section below needs to be updated similarly:
https://github.com/sigp/lighthouse/blob/4273004bd905341f6a3aa3ecbe203ce4b66e6497/beacon_node/lighthouse_network/src/types/pubsub.rs#L267-L282

because `LightClientOptimisticUpdate` and `LightClientFinalityUpdate` aren't superstructs, you'll have to implement a function to manually decode the inner `LightClientHeader` using the fork name, a similar example here:

https://github.com/sigp/lighthouse/blob/a403138ed02f981143f41d53c24e2cae0af07baa/common/eth2/src/types.rs#L1757-L1781

eserilev replied:
ahh makes sense, thanks!

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,143 @@
+    #[superstruct(only(Deneb), partial_getter(rename = "execution_payload_header_deneb"))]
+    pub execution: ExecutionPayloadHeaderDeneb<E>,
+
+    #[ssz(skip_serializing, skip_deserializing)]
```
I think we also need `#[serde(skip)]` here to skip serde.

jimmygchen commented on consensus/types/src/light_client_optimistic_update.rs:
```diff
@@ -38,14 +41,8 @@ impl<T: EthSpec> LightClientOptimisticUpdate<T> {
         chain_spec: &ChainSpec,
```
Looks like we can remove this function too!

jimmygchen commented on consensus/types/src/light_client_update.rs:
```diff
@@ -120,22 +124,40 @@ impl<T: EthSpec> LightClientUpdate<T> {
         let finality_branch = attested_state.compute_merkle_proof(FINALIZED_ROOT_INDEX)?;
+
+        let (attested_header, finalized_header) =
+            match chain_spec.fork_name_at_epoch(beacon_state.slot().epoch(T::slots_per_epoch())) {
```
I think you can just match on `beacon_state` here?

eserilev replied:
i think this one can be deleted too!

eserilev replied:
was under the assumption that the linter would pick up dead code, weird that it didnt pick these ones up

jimmygchen replied:
yeah i think pub functions in lib crates are allowed and not considered dead code.

eserilev replied:
changed to match on beacon_state

eserilev committed:
remove dead code

jimmygchen commented on consensus/types/src/light_client_update.rs:
```diff
@@ -86,79 +91,20 @@ pub struct LightClientUpdate<T: EthSpec> {
-        block: BeaconBlock<T>,
-        attested_state: &mut BeaconState<T>,
-        finalized_block: BeaconBlock<T>,
-    ) -> Result<Self, Error> {
```
oops sorry I think this one might actually be needed for now (used in one of the other light client update PR) - we're only computing and caching `LightClientOptimisticUpdate` and `LightClientFinalityUpdate` in [`LightClientServerCache`](https://github.com/sigp/lighthouse/blob/b23c913605ccb4d565a3f687377954aaac6b7d47/beacon_node/beacon_chain/src/light_client_server_cache.rs#L27) at the moment.

eserilev replied:
brought this back

eserilev committed:
add manual ssz decoding/encoding and add ssz_tests_by_fork macro

eserilev committed:
merge deneb types with tests

eserilev committed:
merge ssz tests, revert code deletion, cleanup

eserilev commented:
@jimmygchen I was able to add the manual ssz encoding/decoding stuff. I also merged your light client ssz test branch here (hope thats ok) and the tests are passing. 

Added a new macro `ssz_tests_by_fork!` which does pretty much the same thing `ssz_tests!` does, just with an additional `ForkName` param. Right now this is used as a unit test for the light client related types specifically for the `Deneb` fork. To enable it for other forks is a bit of extra work, as we'd need a `TestRandom` implementation that also accepts a `ForkName` param.  I started working on a `TestRandomByFork` impl that does just that, but I quickly realized that it was a bit of a bigger change than I initially expected and decided to scrap those changes.

eserilev committed:
move chainspec

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,223 @@
+        let payload = block
+            .message()
+            .execution_payload()?
+            .execution_payload_capella()?
+            .to_owned();
+
+        let header = ExecutionPayloadHeaderCapella::from(&payload);
```
We can remove this clone here:

```suggestion
            .execution_payload_capella()?;

        let header = ExecutionPayloadHeaderCapella::from(payload);
```

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,223 @@
+        let payload = block
+            .message()
+            .execution_payload()?
+            .execution_payload_deneb()?
+            .to_owned();
+
+        let header = ExecutionPayloadHeaderDeneb::from(&payload);
```
```suggestion
            .execution_payload_deneb()?;

        let header = ExecutionPayloadHeaderDeneb::from(payload);
```

jimmygchen commented on consensus/types/src/beacon_block_body.rs:
```diff
@@ -592,6 +600,59 @@ impl<E: EthSpec> From<BeaconBlockBody<E, FullPayload<E>>>
 
+impl<T: EthSpec> BeaconBlockBody<T> {
+    pub fn block_body_merkle_proof(
+        &mut self,
```
This doesn't need to be mutable I think

jimmygchen commented on consensus/types/src/light_client_bootstrap.rs:
```diff
@@ -37,46 +26,59 @@ pub struct LightClientBootstrap<T: EthSpec> {
+
+        let header = decoder
+            .decode_next_with(|bytes| LightClientHeader::from_ssz_bytes(bytes, fork_name))?;
+        let current_sync_committee = decoder.decode_next_with(SyncCommittee::from_ssz_bytes)?;
+        let current_sync_committee_branch =
+            decoder.decode_next_with(FixedVector::from_ssz_bytes)?;
```
nit: can use the shorthand `decode_next` here

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -22,19 +23,19 @@ use test_random_derive::TestRandom;
-#[derive(Debug, Clone, Serialize, Deserialize, Decode, Encode, arbitrary::Arbitrary, PartialEq)]
+#[derive(Debug, Clone, Serialize, Deserialize, arbitrary::Arbitrary, PartialEq)]
 #[serde(untagged)]
+// #[ssz(enum_behaviour = "transparent")]
```
This line can be removed

jimmygchen replied:
~~Oh we actually need this if we're going to derive `Encode`~~
This can be removed now we're manually implementing `Encode`

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -53,6 +54,85 @@ pub struct LightClientHeader<E: EthSpec> {
+impl<E: EthSpec> LightClientHeader<E> {
+    pub fn block_to_light_client_header(
+        block: &SignedBeaconBlock<E>,
+        fork_name: ForkName,
```
nit: we can probably just match on the `SignedBeaconBlock` variant without requiring a `fork_name`

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -53,6 +54,85 @@ pub struct LightClientHeader<E: EthSpec> {
+    }
+}
+
+impl<E: EthSpec> Encode for LightClientHeader<E> {
```
I think we can just derive `Encode` without needing a manual impl?

eserilev replied:
When deriving `Encode` I get a `not all enum variants are variably-sized` error message when running `ssz_tests_by_fork!`. The only work around I've found so far is manually implementing `Encode`

jimmygchen replied:
ah right, the limitation here:
https://github.com/sigp/ethereum_ssz/blob/649ed02ef19c6b2c364d0f4b514cc1c14f6b44d0/ssz_derive/src/lib.rs#L530-L534

The `Altair` variant is fixed sized, so I think you're right that we might have to manually implement this.

jimmygchen commented on testing/ef_tests/check_all_files_accessed.py:
```diff
@@ -27,20 +27,6 @@
     # light_client
     "tests/.*/.*/light_client",
-    # LightClientStore
-    "tests/.*/.*/ssz_static/LightClientStore",
```
I think we'll need to keep `LightClientStore` and `LightClientSnapshot` here as we don't support these types.

jimmygchen commented on testing/ef_tests/tests/tests.rs:
```diff
@@ -285,6 +284,46 @@ mod ssz_static {
 
+    #[test]
+    fn light_client_bootstrap() {
+        SszStaticHandler::<LightClientBootstrap<MinimalEthSpec>, MinimalEthSpec>::altair_only()
```
`altair_and_later` for all these tests now we support them.

jimmygchen reviewed (changes requested):
I've left a few more comments, but we're very close! 🎉 

I think this PR would enable serving `LightClientBootstrap`, `LightClientOptimisticUpdate` and `LightClientOptimisticUpdate` over the REST APIs, and receiving light client gossip updates. It would be worth testing this before we merge.

jimmygchen commented on consensus/types/src/beacon_block_body.rs:
```diff
@@ -592,6 +600,59 @@ impl<E: EthSpec> From<BeaconBlockBody<E, FullPayload<E>>>
+            _ => return Err(Error::IndexNotSupported(generalized_index)),
+        };
+
+        let mut leaves = vec![
+            self.randao_reveal().tree_hash_root(),
+            self.eth1_data().tree_hash_root(),
+            self.graffiti().tree_hash_root(),
+            self.proposer_slashings().tree_hash_root(),
+            self.attester_slashings().tree_hash_root(),
+            self.attestations().tree_hash_root(),
+            self.deposits().tree_hash_root(),
+            self.voluntary_exits().tree_hash_root(),
+        ];
```
FYI @michaelsproul's made some improvements on `tree-states` to make this a bit more maintainable and less error prone:

https://github.com/sigp/lighthouse/blob/928915c71885a2439f8fa3b3a22ef6b39e978339/consensus/types/src/beacon_state.rs#L2200-L2226

it's on the `tree-states` branch though, so ignore this for now, but thought worth mentioning this.

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,223 @@
+}
+
+impl<E: EthSpec> Encode for LightClientHeader<E> {
+    fn is_ssz_fixed_len() -> bool {
```
I think the `is_ssz_fixed_len` and `ssz_bytes_len` implementations may need to account for the fixed sized `Altair` variant?

jimmygchen commented on consensus/types/src/light_client_header.rs:
```diff
@@ -1,26 +1,223 @@
+    }
+
+    fn ssz_bytes_len(&self) -> usize {
+        match self {
```
Would be nice to use some `superstruct` magic to make some of these a bit more maintainable for future forks:

```rust
map_light_client_header_ref!(&'a _, self.to_ref(), |inner, cons| {
    cons(inner);
    inner.ssz_bytes_len()
})
```

eserilev committed:
update ssz tests

eserilev committed:
fmt

dapplion commented:
Would like to see if it's simpler or worse to make `LightClientOptimisticUpdate` a superstruct instead of rolling our own decoding logic. Otherwise looks good, but spec tests for lightclient data structures are breaking

eserilev committed:
light client ssz tests

eserilev committed:
change to superstruct

eserilev commented:
had to make some small tweaks to the ef_test framework to get things working. 

@dapplion @jimmygchen I have a PR here that uses superstruct for all the LightClient types

https://github.com/eserilev/lighthouse/pull/2/files

superstruct is more compatible with the ef_test framework, and removes the need for manual encode/decode implementations. It does make certain parts of the code a bit more verbose, and there might be some additional maintenance work needed for new forks. Lmk what you guys think!

jimmygchen commented:
@eserilev I've reviewed your `superstruct` PR and it looks good. I've added some small comments, but we're _really_ close!

eserilev committed:
changes from feedback

eserilev committed:
linting

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into capella_deneb_light_client_types

eserilev committed:
test fix

eserilev committed:
cleanup

jimmygchen commented on consensus/types/src/beacon_block_body.rs:
```diff
@@ -47,11 +54,12 @@ pub const BLOB_KZG_COMMITMENTS_INDEX: usize = 11;
     partial_getter_error(ty = "Error", expr = "Error::IncorrectStateVariant")
 )]
-#[derive(Debug, Clone, Serialize, Deserialize, Derivative, arbitrary::Arbitrary)]
+#[derive(Debug, Clone, Serialize, Deserialize, Derivative, TreeHash, arbitrary::Arbitrary)]
```
I think this `TreeHash` derive here may be unnecessary

jimmygchen replied:
removed in https://github.com/sigp/lighthouse/pull/4946/commits/d314d12a0f23b65df1d953e18778fdf59d327bfb

jimmygchen commented on consensus/types/src/beacon_block_header.rs:
```diff
@@ -13,6 +13,7 @@ use tree_hash_derive::TreeHash;
 #[derive(
     arbitrary::Arbitrary,
     Debug,
+    Default,
```
This doesn't seem to be used

jimmygchen replied:
removed in https://github.com/sigp/lighthouse/pull/4946/commits/d314d12a0f23b65df1d953e18778fdf59d327bfb

jimmygchen committed:
Remove unused `derive`.

jimmygchen reviewed (approved):
Looking good! 
I've tried running this branch on mainnet and able to get capella light client data using the REST endpoints (light client bootstrap, optimistic updates, finality updats) 🎉 Nice work!

Let's get this merged soon (wait a few days since we've just released)

jimmygchen commented:
btw @dapplion the optimistic updates & finality caching seems to be working great, response times were pretty much instant 🥳

realbigsean commented:
@Mergifyio queue

eserilev committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into capella_deneb_light_client_types

eserilev committed:
beta compiler fix

eserilev committed:
merge

eserilev commented:
resolved a beta compiler issue

realbigsean commented:
@Mergifyio requeue

realbigsean commented:
@Mergifyio queue

jimmygchen commented:
@mergify queue

jimmygchen commented:
@mergify requeue

jimmygchen commented:
@mergify requeue
